File: frontend/src/types/threatIntel.ts
export type IndicatorType =
  | "IP"
  | "DOMAIN"
  | "HASH";

export type ThreatSeverity =
  | "CLEAN"
  | "LOW"
  | "MEDIUM"
  | "HIGH"
  | "CRITICAL";

export interface ThreatIndicator {
  id: string;
  indicator: string;
  type: IndicatorType;
  severity: ThreatSeverity;
  reputation_score: number;
  malicious: boolean;
  confidence: number;
  country?: string;
  provider?: string;
  first_seen?: string;
  last_seen?: string;
  tags: string[];
  description: string;
}

export interface ThreatLookupResponse {
  indicator: ThreatIndicator;
}

File: backend/app/schemas/threat_intel.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator

IndicatorType = Literal["IP", "DOMAIN", "HASH"]
ThreatSeverity = Literal["CLEAN", "LOW", "MEDIUM", "HIGH", "CRITICAL"]

class ThreatLookupRequest(BaseModel):
    indicator: str = Field(min_length=1, max_length=255)
    type: IndicatorType

    @field_validator("indicator")
    @classmethod
    def normalize_indicator(cls, value: str) -> str:
        return value.strip()

class ThreatIndicator(BaseModel):
    id: str
    indicator: str
    type: IndicatorType
    severity: ThreatSeverity
    reputation_score: float
    malicious: bool
    confidence: float
    country: str | None = None
    provider: str
    first_seen: str | None = None
    last_seen: str | None = None
    tags: list[str] = []
    description: str

class ThreatLookupResponse(BaseModel):
    indicator: ThreatIndicator
